python/dgl/nn/pytorch/hetero.py

```diff
@@ -250,3 +251,137 @@ def get_aggregate_fn(agg):
+                              ('user', 'follows', 'user') : ...})
+    >>> print(hg.num_nodes('user'), hg.num_nodes('movie'))
+    300 1000
+    >>> hg.nodes['user'].data['feat'] = torch.randn(hg.num_nodes('user'), 32)
+    >>> hg.nodes['movie'].data['feat'] = torch.randn(hg.num_nodes('movie'), 64)
+    >>> layer = dgl.nn.HeteroLinearLayer(hg, 100, 'feat')
+    >>> g, feat = layer(hg)
+    >>> print(g.num_nodes())
+    1300
+    >>> print(feat.shape)
+    (1300, 100)
+    """
+    def __init__(self, hg, out_size, feat_name):
+        super(HeteroLinearLayer, self).__init__()
+        self.feat_name = feat_name
+        self.linears = nn.ModuleDict()
+        for ntype in hg.ntypes:
+            linear = nn.Linear(hg.nodes[ntype].data[feat_name].shape[1], out_size)
+            self.linears[ntype] = linear
+    
+    def forward(self, hg):
+        """Forward function
+        
+        Parameters
+        ----------
+        hg : DGLGraph
+            The input heterogeneous DGL graph.
+
+        Returns
+        -------
+        g : DGLGraph
+            The homogenized DGL graph.
+        feat : Tensor
+            Transformed node features.
+        """
+        feat = th.tensor([])
+        for ntype in hg.ntypes:
+            features = self.linears[ntype](hg.nodes[ntype].data[self.feat_name])
+            feat = th.cat((feat, features))
+        g = to_homogeneous(hg)   
+        
+        return g, feat          
+    
+class HeteroEmbedding(nn.Module):
+    """Create node embeddings for each node type and return a homogeneous
+    graph representation.
+
+    The underlying implementation invokes dgl.to_homogeneous. Therefore,
+    the returned graph stores the following extra attributes:
+    
+      - ret_g.ndata[dgl.NID]: the original node IDs.
+      - ret_g.edata[dgl.EID]: the original edge IDs.
+      - ret_g.ndata[dgl.NTYPE]: the type ID of each node.
+      - ret_g.edata[dgl.ETYPE]: the type ID of each edge.
+
+    Parameters
+    ----------
+    hg : DGLGraph
+        A heterogeneous DGL graph.
+    embed_size : int
+        Node embedding size.
+
+    Examples
+    --------
+    >>> hg = dgl.heterograph({('user', 'rate', 'movie') : ...,
+                              ('user', 'follows', 'user') : ...})
+    >>> print(hg.num_nodes('user'), hg.num_nodes('movie'))
+    300 1000
+    >>> layer = dgl.nn.HeteroEmbedding(hg, 100)
+    >>> g, embed = layer(hg)
+    >>> print(g.num_nodes())
+    1300
+    >>> print(embed.shape)
+    (1300, 100)
+    >>> print(embed.requires_grad)
+    True
+    """
+    def __init__(self, hg, embed_size):
+        super(HeteroEmbedding, self).__init__()
+        
+        self.embed_size = embed_size
+        self.embed = {}
+        for ntype in hg.ntypes:
+            nodes = hg.num_nodes(ntype = ntype)
+            self.embed[ntype] = nn.Parameter(th.FloatTensor(nodes, self.embed_size))
+            nn.init.xavier_uniform_(self.embed[ntype], gain=nn.init.calculate_gain('relu'))
```

Comment:
Do we still need to initialize node embeddings per node type? Btw, `hg.num_nodes()` returns the total number of nodes across all node types.

Reply:
This is not very clear to me, because I have seen many of them with the initialization operation. In hg.num_nodes() we did get the total number of nodes across all node types, but if we specify the node type, we can get the number of nodes of the corresponding type.

Reply:
What do you mean by "because I have seen many of them with the initialization operation"? Initializing node embeddings per node type and then concatenating them should be equivalent to initializing node embeddings of all nodes at once.

Reply:
OK, I'm thinking too complicated. In the end we get homogeneous graphs, so we do not need the dictionary. And I will update this.